core/Tracker/FingerprintSalt.php

```diff
@@ -0,0 +1,88 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+
+namespace Piwik\Tracker;
+
+use Piwik\Common;
+use Piwik\Date;
+use Piwik\Exception\InvalidRequestParameterException;
+use Piwik\Exception\UnexpectedWebsiteFoundException;
+use Piwik\Option;
+use Piwik\Piwik;
+use Piwik\SettingsServer;
+use Piwik\Site;
+use Piwik\Db as PiwikDb;
+
+class FingerprintSalt
+{
+    const OPTION_PREFIX = 'fingerprint_salt_';
+    const DELETE_FINGERPRINT_OLDER_THAN_SECONDS = 432000; // 5 days in seconds
+
+    public function generateSalt()
+    {
+        return Common::getRandomString(32);
+    }
+
+    public function deleteOldSalts()
+    {
+        // we want to make sure to delete salts that were created more than three days ago as they are likely not in
+        // use anymore. We should delete them to ensure the fingerprint is truly random for each day because if we used
+        // eg the regular salt then it would technically still be possible to try and regenerate the fingerprint based
+        // on certain information.
+        // Typically, only the salts for today and yesterday are used. However, if someone was to import historical data
+        // for the same day and this takes more than five days, then it could technically happen that we delete a
+        // fingerprint that is still in use now and as such after deletion a few visitors would have a new configId
+        // within one visit and such a new visit would be created. That should be very much edge case though.
+        $deleteSaltsCreatedBefore = Date::getNowTimestamp() - self::DELETE_FINGERPRINT_OLDER_THAN_SECONDS;
+        $options = Option::getLike(self::OPTION_PREFIX . '%');
+        $deleted = array();
+        foreach ($options as $name => $value) {
+            $value = $this->decode($value);
+            if (empty($value['time']) || $value['time'] < $deleteSaltsCreatedBefore) {
+                Option::delete($name);
+                $deleted[] = $name;
+            }
+        }
+
+        return $deleted;
+    }
+
+    public function getDateString(Date $date, $timezone)
+    {
+        $dateString = Date::factory($date->getTimestampUTC(), $timezone)->toString();
```

Comment:
not 100% sure if this is correct? Had initially `Date::factory()->setTimezone($timezone)` but then noticed in tests it might not return correct values unless I'm expecting the wrong values. In ArchiveProcessor we're using for example `setTimezone`: https://github.com/matomo-org/matomo/blob/3.13.5/core/ArchiveProcessor/Parameters.php#L166 so would have maybe expected `setTimezone` is right but then in CronArchive we're also using it as a parameter https://github.com/matomo-org/matomo/blob/3.13.5/core/CronArchive.php#L1504 ... looking at the test results the parameter seems to give expected result

Reply:
Guess that's something we should probably check again with #13829